[Test] Replace 0.0.0 version with "detached" property on ElasticsearchDistibution

jozala opened the pull request:
Introduce a new property to indicate if the version is a custom build not tied to any release. This was previously handled by specifying version as 0.0.0, but caused problems when comparing versions (e.g., to decide on configuration).

In the tests, which support the `bcUpgradeTest` task it means that they need to:
- apply the `elasticsearch.bc-upgrade-test` (this was already required),
- configure the version of the cluster/node using the ElasticsearchCluster JUnit rule (It was already required, but after this change it can be read from `tests.bwc.main.version` or `tests.old_cluster_version` since both are the same. Previously, sometimes version had to be applied from `tests.old_cluster_version` which could be `0.0.0` to be able to support special cases.)
- set `detachedVersion` to `true` using the ElasticsearchCluster JUnit rule when the cluster should resolve to the version not tied to any released version nor the current HEAD (This is new. This setting is currently tied to the existence of the `tests.bwc.refspec.main` system property).

I haven't found any of the legacy rest/yaml tests supporting `bcUpgradeTest` (none applies `elasticsearch.bc-upgrade-test` plugin), so I haven't been implementing any support for detached version for the legacy test framework.

jozala committed:
[Test] Replace 0.0.0 version with "detached" property on
ElasticsearchDistibution for tests

Introduce a new property to indicate if the version is a custom build
not tied to any release. This was previously handled by specifying
version as 0.0.0, but caused problems when comparing versions (e.g., to
decide on configuration).

jozala committed:
Change test clusters configuration to support bcUpgradeTest explicitly

Added
`.detachedVersion(System.getProperty("tests.bwc.refspec.main") != null)`
to all the tests which apply 'elasticsearch.bc-upgrade-test' plugin to
make them properly resolve distribution (previously configured by
specifying `0.0.0` version).

jozala committed:
Change version handling in the tests

In the bcUpgradeTests these tests were previously skipped because
version was set to `0.0.0` which was outside the supported range.
Now, when `tests.old_cluster_version` has correct version they run, but
fail just because `9.2.0-SNAPSHOT` was considered unsupported version
format with the "SNAPSHOT".

jozala committed:
Use distribution version in the detached DistributionResolver

Update logic to use version from distribution. It no longer needs to use
 overridden version in system property as it should be the same as in
 the distribution.

jozala committed:
Reset the detachedVersion flag on upgrading cluster

This is needed so that the cluster is upgraded to the actual specified
version and not keep using the same detached version as before.
It has a limitation of not being able to upgrade to the detached
version, but there is no such test case currently.

jozala committed:
Allow version 0.0.0 for tests

Need to keep allowing that for serverless tests that doesn't know the
version for the detached version distribution when registering the BwC
test tasks.

jozala committed:
Expect null version in MixedClusterEsqlSpecIT

This is needed explicitly now for serverless tests. Previously it was
also null but was implicit.

Commit:
[CI] Auto commit changes from spotless

jozala commented on qa/full-cluster-restart/build.gradle:
```diff
@@ -23,10 +23,18 @@ buildParams.bwcVersions.withIndexCompatible { bwcVersion, baseName ->
-  usesBwcDistribution(Version.fromString("0.0.0"))
-  systemProperty("tests.old_cluster_version", "0.0.0")
-  onlyIf("tests.bwc.main.version system property exists") { System.getProperty("tests.bwc.main.version") != null }
+  doFirst {
+    if (System.getProperty("tests.bwc.refspec.main") == null) {
+      throw new GradleException("You must set the `tests.bwc.refspec.main` system property to run luceneBwcTest")
+    }
+    if (System.getProperty("tests.bwc.main.version") == null) {
+      throw new GradleException("You must set the `tests.bwc.main.version` system property to run luceneBwcTest")
+    }
+  }
+  if (System.getProperty("tests.bwc.refspec.main") != null && System.getProperty("tests.bwc.main.version") != null) {
+    usesBwcDistributionFromRef(System.getProperty("tests.bwc.refspec.main"), Version.fromString(System.getProperty("tests.bwc.main.version")))
+    systemProperty("tests.old_cluster_version", System.getProperty("tests.bwc.main.version"))
+  }
```
Same logic change as in [/elasticsearch.bc-upgrade-test.gradle](https://github.com/elastic/elasticsearch/pull/134584/files#diff-8238678af910cb4e0c291c57e05b4c036935b10b4704909946b37422819af563)

mark-vieira commented on build-tools/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java:
```diff
@@ -91,6 +93,19 @@ public void setVersion(String version) {
         this.version.set(version);
     }
 
+    /**
+     * Informs if the version is not tied to any Elasticsearch release and is a custom build.
+     * This is true when the distribution is not from HEAD but also not any known released version.
+     * In that case the detached source build needs to be prepared by `usedBwcDistributionFromRef(ref, version)`.
+     */
+    public boolean isDetachedVersion() {
+        return detachedVersion.get();
+    }
+
+    public void setDetachedVersion(boolean detachedVersion) {
+        this.detachedVersion.set(detachedVersion);
+    }
+
```
I like the term `detatched`. I certainly can't think of a better one and I agree "snapshot" is both overloaded and not really appropriate here.

mark-vieira commented on test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/DefaultLocalClusterHandle.java:
```diff
@@ -167,13 +167,15 @@ public void upgradeNodeToVersion(int index, Version version) {
         Node node = nodes.get(index);
         node.stop(false);
         LOGGER.info("Upgrading node '{}' to version {}", node.getName(), version);
+        node.getSpec().setDetachedVersion(false);
```
Yeah, this separation is what makes this somewhat awkward. Which is why I originally envisioned the notion of "detached" here to be captured by `Version` itself. All the test clusters stuff uses it's own `Version` class, so we could add a `detached` property to it. It would save us having to muck with all these method signatures and API that already accept a `Version` while still being able to pass that knowledge down to things that care about that.

So how about we add `detached` to `Version`. We'll then add some factory methods that take the additional boolean argument. We can do the same on `LocalSpecBuilder` by adding a `version(String, boolean)` method. If unspecified the default will be to set `detached = false`. That way this all "just works". We just look at the `detached` property of the passed `Version` to figure out what to do. This also allows us in the future to have tests that upgrade _to_ detached versions as well. As this is written that wouldn't be possible.

mark-vieira reviewed (changes requested):
See comments regarding modeling `detached` as a property of `Version` rather than a separate configurable property on clusters themselves. Otherwise these changes look great.

jozala committed:
Merge branch 'main' into detached-property-on-esdistro

jozala committed:
Refactor ElasticsearchCluster detached version handling

Updated the ElasticsearchCluster initialization to use a single method
for setting the version and detached version flag. It also integrates
the detached property into Version to make it transparent in case the
detached functionality is not used in the test.

jozala committed:
Do not check for `0.0.0` version for JDK-OS compatibility

This check in no longer needed, because the version should be now set to
 the real value instead of `0.0.0`, even in the bcUpgradeTest task

jozala committed:
Spotless fixes

jozala committed:
Make Version constructor protected to be able to use detached from Serverless

breskeby commented on build-tools-internal/src/main/groovy/elasticsearch.bc-upgrade-test.gradle:
```diff
@@ -11,8 +11,16 @@ import org.elasticsearch.gradle.Version
-  systemProperty("tests.old_cluster_version", "0.0.0")
-  onlyIf("tests.bwc.main.version system property exists") { System.getProperty("tests.bwc.main.version") != null }
+  doFirst {
+    if (System.getProperty("tests.bwc.refspec.main") == null) {
```
Can't we just move the check for the required system properties to the task configuration phase? would we ever wanna call those tasks without these params?

breskeby replied:
lets do the same in the bcupgrade script plugin

jozala replied:
Thanks for looking into that Rene.
I've changed that now in both places to use `whenReady` instead of `doFirst`.

jozala committed:
Verify the availability of the required system property with Gradle whenReady

Using `project.gradle.taskGraph.whenReady` to check the required system properties as soon as possible.

mark-vieira reviewed (approved):
Two minor comments but otherwise LGTM.

jozala committed:
Merge branch 'main' into detached-property-on-esdistro

jozala committed:
Make detached part of the equals and hashcode methods for Version

jozala committed:
Revert to use Pair for Distribution key

jozala commented on test/test-clusters/src/main/java/org/elasticsearch/test/cluster/local/AbstractLocalClusterFactory.java:
```diff
@@ -397,8 +397,8 @@ private void copyExtraJarFiles() {
             return TEST_DISTRIBUTIONS.computeIfAbsent(
-                Pair.of(spec.getVersion(), spec.getDistributionType()),
-                key -> distributionResolver.resolve(key.left, key.right)
+                new DistributionKey(spec.getVersion(), spec.getDistributionType()),
```
I've reverted that change to use Pair again.

jozala commented on test/test-clusters/src/main/java/org/elasticsearch/test/cluster/util/Version.java:
```diff
@@ -33,6 +33,7 @@ public class Version implements Comparable<Version>, Serializable {
     private final int revision;
     private final int id;
     private final String qualifier;
+    private final boolean detached;
```
Good point. Done.

jozala committed:
Merge branch 'main' into detached-property-on-esdistro

jozala committed:
Version `0.0.0` exception reverted

Needs to stay there, because of the serverless tests.

jozala commented:
This doesn't need to be backported to 8.19 and 8.18 because `bcUpgradeTest` and version `0.0.0` was never introduced on these branches.
